File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        res = []
        FrequencyMap = {} # number -> frequency count

        for n in nums:
            FrequencyMap[n] = 1 + FrequencyMap.get(n,0)
        
        bucket =[[] for _ in range(len(nums)+1)]

        for n,fcount in FrequencyMap.items():
            bucket[fcount].append(n) 
        
        for i in range(len(nums) , 0 , -1):
            for num in bucket[i]:
                res.append(num)

                if k == len(res):
                    return res
            




        
        
        

File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        // USING HASHMAP AND PRIORITY QUEUE 

        HashMap<Integer , Integer> FrequencyCount = new HashMap<>();

        for( int i = 0 ; i < nums.length ; i++)
        {
            FrequencyCount.put(nums[i] , FrequencyCount.getOrDefault(nums[i],0) + 1);
        }

        PriorityQueue<int[]> maxHeap = new PriorityQueue<>((a,b) -> b[1] - a[1]);
        for(Map.Entry<Integer,Integer> entry : FrequencyCount.entrySet()){
            maxHeap.add(new int[]{entry.getKey(), entry.getValue()});
        }
        int[] res = new int[k];

        for(int i = 0; i < k; i++){
            res[i] = maxHeap.poll()[0];
        }

        return res;


        
    }
}